Implements time stepping in MBP

amcastro-tri opened the pull request:
Mostly a systems framework exercise PR. Thus far contact is not supported in the time stepping MBP.
This is a required step for time stepping methods for contact resolution.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8879)
<!-- Reviewable:end -->

amcastro-tri commented:
Unused parameter flake. Will fix shortly.

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

As always, a pleasure reading your code. Just some cosmetic stuff.

---

Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[multibody/multibody_tree/multibody_tree_context.h, line 53 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN8EvZETdMCZJRctTy:-LDN8EvZETdMCZJRctTz:beaxet9) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_tree_context.h#L53)):*
> ```objc
>   /// @param[in] discrete_state
>   ///   `true` if the state is to be stored as discrete state. Otherwise the
>   ///   state is stored as continuous state.
> ```

FYI Meta: While generally, style is to *exclude* the "otherwise" clause, in this case it might be justified. I merely point it out to you. If "not discrete" is blindingly obviously "continuous", then you might consider dropping the second clause. This applies to all of the documentation in this regard. It is worth noting that the `MultibodyPlant` does *not* spell out the otherwise.

---

*[multibody/multibody_tree/multibody_tree_context.h, line 101 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN8yrF0fwE1aRGcz2Z:-LDN8yrF0fwE1aRGcz2_:b-ycjqqx) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_tree_context.h#L101)):*
> ```objc
>   /// Returns `true` if the state is discrete and `false` if the state is
>   /// continuous.
>   bool is_state_discrete() const {
> ```

FYI This name *might* be slightly redundant. Do you need to distinguish other kinds of discreteness in this context subclass? If not, then simply, `is_discrete` would be sufficient.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 163 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMvYyAAvV7rDJi6lpo:-LDMvYyAAvV7rDJi6lpp:bdabd9s) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L163)):*
> ```objc
>   /// @param[in] time_step
>   ///   An optional parameter indicating whether `this` plant is modeled as a
>   ///   continuous system (`time_step = 0`) or, as a discrete system with
> ```

BTW No comma (and even if one were to argue that there should be a comma, it should precede the conjunction "or").

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 165 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMwXkx2i-EF3FsQiYw:-LDMwXkx2i-EF3FsQiYx:b-tzselz) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L165)):*
> ```objc
>   ///   continuous system (`time_step = 0`) or, as a discrete system with
>   ///   periodic updates of period `time_step > 0`. @default 0.0.
>   /// @throws std::exception if `time_step` is negative.
> ```

FYI As a design decision, you could basically say positive --> time stepping, everything else --> continuous. Is there *particular* value in having some values throw an exception?

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 754 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMwtOwDv6relTL9Gb3:-LDMwtOwDv6relTL9Gb4:bsxgguk) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L754)):*
> ```objc
>   /// The time step (or period) used to model `this` plant as a discrete system
>   /// with periodic updates. Returns 0 (zero) if the plant is modeled as a
>   /// continuous system.
> ```

BTW Probably worth explicitly referring to the documentation of the constructor for how to *interpret* the returned value.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 943 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMxKbb7EXnHhW-1vVs:-LDMxKbb7EXnHhW-1vVt:binod83) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L943)):*
> ```objc
>   //   vⁿ⁺¹ = vⁿ + dt v̇ⁿ
>   //   qⁿ⁺¹ = qⁿ + dt N(qⁿ) vⁿ⁺¹
>   // this semi-explicit update is symplectic, which for Hamiltonian system has
> ```

BTW s/this/This

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 946 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMxPIX5YhO9gdRC8nl:-LDMxPIX5YhO9gdRC8nm:b-qh0sar) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L946)):*
> ```objc
>   // the nice property of nearly conserving energy (in many cases we can write a
>   // "modified energy functional" which can be shown to be exactly conserved and
>   // to be within O(dt) of the real energy of the mechanical system.
> ```

BTW Missing closing parenthesis.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 499 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMzKFoBWXshhY3B2ua:-LDMzKFoBWXshhY3B2ub:b-vf92qx) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L499)):*
> ```cpp
>     drake::systems::DiscreteValues<T>* updates) const {
>   // If plant state is continuous, no discrete state to update.
>   if (!is_discrete()) return;
> ```

I shouldn't think this test should be necessary. As I understand it, this plant's discrete update event handler should *only* get invoked if it has explicitly asked for it to be invoked (i.e., by registering the appropriate event). So, it seems the plant should only register that event if it is discrete (as you have done below). So, doing the test here should be redundant.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 510 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN-nhp-itzGYwHCTew:-LDN-nhqEDkcfyEQZQTW:b-6i6xcr) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L510)):*
> ```cpp
>   const int nv = this->num_velocities();
> 
>   // Get the system state as a raw Eigen vectors
> ```

BTW Remove "a" (as in "...as raw Eigen vectors")

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 521 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN0pQ28Qb4OiadWiLX:-LDN0pQ28Qb4OiadWiLY:b-249e9g) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L521)):*
> ```cpp
>   auto M0_ldlt = M0.ldlt();
> 
>   // Forces at the previous time step.
> ```

BTW This comment *may* be a bit misleading. This isn't the forces, it's the container for the forces (which are actually computed on line 529).

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 529 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN0NR61BbexzfPK-Sn:-LDN0NR74mnnpTK-PN7W:biff6ib) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L529)):*
> ```cpp
> 
>   // Compute forces applied through force elements. This effectively resets
>   // the forces to zero and adds in contributions due to force elements.
> ```

BTW This comment confused me.

As near as I can tell, `forces0` is already zeroed out (as the result of the constructor). I'd recommend cutting out the second sentence entirely unless there is some import nuance that you still intend.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 563 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN1V-S09y95euCvUnT:-LDN1V-S09y95euCvUnU:b-7xl1ad) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L563)):*
> ```cpp
>   std::vector<SpatialForce<T>>& F_BBo_W_array = forces0.mutable_body_forces();
> 
>   // With vdot = 0, this computes (includes normal forces):
> ```

BTW "...this computes (__*including*__ normal forces):"

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 570 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN1bzg8d9eK-KAs9oW:-LDN1bzg8d9eK-KAs9oX:b4m3hkd) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L570)):*
> ```cpp
>       F_BBo_W_array, minus_tau,
>       &A_WB_array,
>       &F_BBo_W_array, /* Notice these arrays gets overwritten on output. */
> ```

BTW s/Notice/Note:

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 579 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN1sBE1571nfoA2gyN:-LDN1sBE1571nfoA2gyO:b-29bdp1) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L579)):*
> ```cpp
>   VectorX<T> q_next = q0 + dt * qdot_next;
> 
>   // q_next = q0 + dt * qdot_next.
> ```

BTW This comment is lost. It's exactly the line of code two lines up.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 590 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN27ek6UJtN7rv5xV1:-LDN27ek6UJtN7rv5xV2:b76ak1t) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L590)):*
> ```cpp
>     const Eigen::Ref<const VectorX<T>>& qdot,
>     systems::VectorBase<T>* generalized_velocity) const {
>   if (is_discrete()) return;
> ```

BTW You have a number of functions that you've turned into no-ops for the discrete system. I'd recommend *documenting* somewhere the basis for this choice. I'm left wondering if they *should* be called in discrete systems at all, and, if not, that should be a hard error.

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 153 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN66jBDv0lcb0TwwXH:-LDN2wyN9t0c1ru38Wu7:bf9bpj5) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L153)):*
> ```cpp
>     // Make a non-finalized plant so that we can tests methods with pre/post
>     // Finalize() conditions.
>     const std::string full_name = FindResourceOrThrow(
> ```

FYI Not clear why this is part of this PR. Why couldn't `MakeAcrobotPlant` simply have been extended to include a time-stepping parameter?

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 258 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN6BT39I3EGWBCQZxq:-LDN6BT39I3EGWBCQZxr:b-dg8iac) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L258)):*
> ```cpp
>   }
> 
>   void VerifyDoCalcDiscreteVariableUpdates(double theta1, double theta2,
> ```

BTW Please put a comment that this verifies the discrete system against the continuous system. Also address the reason for using the continuous time derivatives instead of the reference benchmark against which the derivatives are verified. (In other words, document *what* is being verified, exactly).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879#-:-LDN9KjiAb-IuNeqw6zR:b-yud829)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform pass done. There appears to be a conflict between MultibodyContext's allocation of state variables and MultibodyPlant's declarations of same, please see below.

---

Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[multibody/multibody_tree/mobilizer_impl.h, line 88 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN3NFo3u7XgQkWrep-:-LDN3NFo3u7XgQkWrep0:b-6n0j8a) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/mobilizer_impl.h#L88)):*
> ```objc
>   /// Eigen::VectorBlock<VectorX<T>>.
>   Eigen::VectorBlock<VectorX<T>> get_mutable_state_vector(
>       const systems::Context<T>& context, systems::State<T>* state) const {
> ```

BTW Do you expect the State here to be part of the same Context you're passing in? If so it would be better just to take a mutable Context as the sole parameter.

---

*[multibody/multibody_tree/multibody_tree_context.h, line 81 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN4fZK0Corh1tKMca0:-LDN4fZK0Corh1tKMca1:b-2he2fo) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_tree_context.h#L81)):*
> ```objc
>     } else {
>       auto xd = std::make_unique<DiscreteValues<T>>(
>           std::make_unique<BasicVector<T>>(num_states));
> ```

This appears to be unused? Is a set_discrete_state() missing? If so I'm surprised this passed unit tests -- please make sure there is something downstream that cares!

---

*[multibody/multibody_tree/multibody_tree_context.h, line 158 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN61qG6-2qaRUOvKd2:-LDN61qG6-2qaRUOvKd3:b-eufxh2) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_tree_context.h#L158)):*
> ```objc
>     // TODO(amcastro-tri): make use of VectorBase::get_contiguous_vector() once
>     // PR #6049 gets merged.
>     Eigen::VectorBlock<const VectorX<T>> xc = get_state_vector();
> ```

BTW this shouldn't be xc now since it might be xd. Maybe just x? Or xₖ for kinematic states?

---

*[multibody/multibody_tree/multibody_tree_context.h, line 220 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMznXkAWr1sxqD0fP3:-LDMznXl8Sp1txhYZpNH:b9r81sz) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_tree_context.h#L220)):*
> ```objc
>   const MultibodyTreeTopology topology_;
> 
>   // If `true`, this context stores a discrete state. If `false` the sate is
> ```

BTW sate -> state

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 165 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN6sBE2S6QNQJV1TFl:-LDN6sBE2S6QNQJV1TFm:b-u5031e) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L165)):*
> ```objc
>   ///   continuous system (`time_step = 0`) or, as a discrete system with
>   ///   periodic updates of period `time_step > 0`. @default 0.0.
>   /// @throws std::exception if `time_step` is negative.
> ```

BTW this is not correct -- it is using DRAKE_DEMAND so aborts instead. 

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 747 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN07neDLIbl3vh_NCk:-LDN07neDLIbl3vh_NCl:b-6sohai) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L747)):*
> ```objc
> 
>   /// Returns `true` if `this` plant is modeled as a discrete system with
>   /// periodic updates of period equal to time_step().
> ```

BTW from the comment I thought I'd have to provide the time_step() here for comparison. Better here just to say "Returns true if this plant is modeled as a discrete system."

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 943 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN0pwz0riOVo4FRd6A:-LDN0pwz0riOVo4FRd6B:b-iqwozr) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L943)):*
> ```objc
>   //   vⁿ⁺¹ = vⁿ + dt v̇ⁿ
>   //   qⁿ⁺¹ = qⁿ + dt N(qⁿ) vⁿ⁺¹
>   // this semi-explicit update is symplectic, which for Hamiltonian system has
> ```

Semi-_explicit_ is not symplectic unless there are no velocity-dependent forces (including Coriolis forces). The semi-_implicit_ form (a.k.a. symplectic Euler) would be:
```
  //   vⁿ⁺¹ = vⁿ + dt v̇ⁿ⁺¹
  //   qⁿ⁺¹ = qⁿ + dt N(qⁿ) vⁿ⁺¹
```
You could say that in many cases the semi-explicit form inherits nice properties from semi-implicit, in common cases where the velocity-based forces are small compared to other forces.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 53 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN72L2A-xZfT42-DzS:-LDN72L2A-xZfT42-DzT:b-d4b58w) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L53)):*
> ```cpp
>         drake::multibody::multibody_plant::MultibodyPlant>()),
>     time_step_(time_step) {
>   DRAKE_DEMAND(time_step >= 0);
> ```

BTW inconsistent with the API comment. You could use DRAKE_THROW_UNLESS to generate an exception rather than an abort, or leave this and change the comment just to say time_step ≥ 0 is a precondition.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 636 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN9AcV-E9jz_yaBPpL:-LDN9AcV-E9jz_yaBPpM:bafzm5x) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L636)):*
> ```cpp
>         model_->num_positions(),
>         model_->num_velocities(), 0 /* num_z */);
>   }
> ```

The above code appears to duplicate the state allocation being done in MultibodyContext's constructor. LeafContext automatically adds continuous and discrete state from these declarations, but I think that is getting done on top of state allocations done in MultibodyContext. I speculate that that is why this works even though the discrete state was never allocated in MultibodyContext.

Better to do it just here and not attempt to allocate states in MBContext, I think.

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 276 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDNAfPo0ehkzgt2F5RI:-LDNAfPo0ehkzgt2F5RJ:bmugehw) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L276)):*
> ```cpp
>     // to whether the state is discrete or continuous. Therefore, we are allowed
>     // to using the same modeling elements to set both `context` and
>     // `discrete_context`.
> ```

BTW nice!

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 313 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDNALQF0UuZYhhVUEec:-LDNALQF0UuZYhhVUEed:bp18zq2) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L313)):*
> ```cpp
>   // The model plant:
>   MultibodyPlant<double>* plant_{nullptr};
>   std::unique_ptr<MultibodyPlant<double>> discrete_plant_;
> ```

BTW why is plant_ a raw pointer but discrete_plant_ a unique_ptr? Can you make them the same?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879#-:-LDNBE721ZvP-GOsrXNr:b-lup6ng)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[multibody/multibody_tree/mobilizer_impl.h, line 88 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN3NFo3u7XgQkWrep-:-LDNOUWHEUK4ssqjDj-1:b-f1j6tq) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/mobilizer_impl.h#L88)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW Do you expect the State here to be part of the same Context you're passing in? If so it would be better just to take a mutable Context as the sole parameter.
</blockquote></details>

BTW an entire train of calls and API driven by the need to make `MBP::SetDefaultState(context, *sate)` to work. Nothing I can easily fix right now. Probably if there was no distinction in the state between continuous and discrete it'd be better?

---

*[multibody/multibody_tree/multibody_tree_context.h, line 81 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN4fZK0Corh1tKMca0:-LDNPKDj9HKF3HImrjOp:b-xpfm2o) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_tree_context.h#L81)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This appears to be unused? Is a set_discrete_state() missing? If so I'm surprised this passed unit tests -- please make sure there is something downstream that cares!
</blockquote></details>

True. It might only be needed for a standalone MBT application with discrete state. Otherwise the plant's `Declare` statements do everything for us. This clearly needs fixing but it'll take me a while, not in this PR. I added a big TODO explaining the situation.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 165 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMwXkx2i-EF3FsQiYw:-LDNR4NlEV3Il_iKG9hk:b-slxeuc) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L165)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI As a design decision, you could basically say positive --> time stepping, everything else --> continuous. Is there *particular* value in having some values throw an exception?
</blockquote></details>

Not really, just a bid odd to pass a "negative" period. Zero, the limit case, still is ok. Stylistics I guess.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 165 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN6sBE2S6QNQJV1TFl:-LDNRGhK7a6XblF4Bn7m:b-k2rwtz) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L165)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW this is not correct -- it is using DRAKE_DEMAND so aborts instead. 
</blockquote></details>

Good catch. Fixed.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.h, line 943 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN0pwz0riOVo4FRd6A:-LDNTizo58KQQkPnV5gu:bz29jbv) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.h#L943)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Semi-_explicit_ is not symplectic unless there are no velocity-dependent forces (including Coriolis forces). The semi-_implicit_ form (a.k.a. symplectic Euler) would be:
```
  //   vⁿ⁺¹ = vⁿ + dt v̇ⁿ⁺¹
  //   qⁿ⁺¹ = qⁿ + dt N(qⁿ) vⁿ⁺¹
```
You could say that in many cases the semi-explicit form inherits nice properties from semi-implicit, in common cases where the velocity-based forces are small compared to other forces.
</blockquote></details>

ah! nice detail, thanks!

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 499 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMzKFoBWXshhY3B2ua:-LDNUCFy0agOZtcgfH4p:brmqc3x) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L499)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I shouldn't think this test should be necessary. As I understand it, this plant's discrete update event handler should *only* get invoked if it has explicitly asked for it to be invoked (i.e., by registering the appropriate event). So, it seems the plant should only register that event if it is discrete (as you have done below). So, doing the test here should be redundant.
</blockquote></details>

Is that true even if we add another discrete update? (I don't know what for, but I just wonder).

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 563 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN1V-S09y95euCvUnT:-LDNUrnh-9BVI-mRLMK8:b-85pi2z) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L563)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW "...this computes (__*including*__ normal forces):"
</blockquote></details>

well.. that's not true in this PR. Leftover from my dev branch in which that is true.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 590 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN27ek6UJtN7rv5xV1:-LDNVJR_BbUE8Hq--cQT:b-8lzpcr) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L590)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW You have a number of functions that you've turned into no-ops for the discrete system. I'd recommend *documenting* somewhere the basis for this choice. I'm left wondering if they *should* be called in discrete systems at all, and, if not, that should be a hard error.
</blockquote></details>

To be honest, I don't know what the system framework does here. I believe RBP, and now MBP, are the only systems we have that can behave in either mode. RBP does perform the check so I assumed Evan did his homework :-). 

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 636 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN9AcV-E9jz_yaBPpL:-LDNVc2o3Q3qpb3Ja09I:b5w3wym) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L636)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

The above code appears to duplicate the state allocation being done in MultibodyContext's constructor. LeafContext automatically adds continuous and discrete state from these declarations, but I think that is getting done on top of state allocations done in MultibodyContext. I speculate that that is why this works even though the discrete state was never allocated in MultibodyContext.

Better to do it just here and not attempt to allocate states in MBContext, I think.
</blockquote></details>

see reply in `mbt_context`

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 153 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN66jBDv0lcb0TwwXH:-LDNVuzl9acQPiNduf7b:bnmm33) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L153)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Not clear why this is part of this PR. Why couldn't `MakeAcrobotPlant` simply have been extended to include a time-stepping parameter?
</blockquote></details>

Just because loading SDF's is much easier. `MakeXXX()` was just the temporary hack. I clean up as I go.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thank you for the prompt review gentleman, back to you.

---

Review status: 7 of 11 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879#-:-LDNYfyF8ae_QQfXHxQL:b4ni69q)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Thanks for the changes and explanations. Platform <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with one question.

---

Reviewed 1 of 4 files at r2.
Review status: 8 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[multibody/multibody_tree/mobilizer_impl.h, line 88 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN3NFo3u7XgQkWrep-:-LDOJfIy5CXkV5IqZ3ij:b-sv37fx) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/mobilizer_impl.h#L88)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW an entire train of calls and API driven by the need to make `MBP::SetDefaultState(context, *sate)` to work. Nothing I can easily fix right now. Probably if there was no distinction in the state between continuous and discrete it'd be better?
</blockquote></details>

Got it. Is there a reason that the default leaf system implementation of SetDefaultState() doesn't work?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879#-:-LDOJuvB-SrSNIlHEjUC:b-z5k7gb)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
I still have two sticking points. Otherwise, great clean up.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 499 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMzKFoBWXshhY3B2ua:-LDOMmQp7MjuEmCI4nJk:bhefaql) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L499)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Is that true even if we add another discrete update? (I don't know what for, but I just wonder).
</blockquote></details>

If you add another type of event that requires discrete update, then you have to differentiate those -- but until that happens, there's no differentiation required. And, again, the only entity that can enforce this method being called on this system is this system.

Right now I don't like the unnecessary code being here -- it suggests it is necessary and that means the code is lying.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 590 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN27ek6UJtN7rv5xV1:-LDON3Du8xOzgl3Kh40z:blnh7n4) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L590)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

To be honest, I don't know what the system framework does here. I believe RBP, and now MBP, are the only systems we have that can behave in either mode. RBP does perform the check so I assumed Evan did his homework :-). 
</blockquote></details>

It should still be documented. If you can get the explanation from Evan and document it, great.  But this is an undocumented pattern and it would be a shame if it is unnecessarily propagated.

---

*[multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc, line 153 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN66jBDv0lcb0TwwXH:-LDONCtHDxe23txqwU9y:b-eied9y) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/test/multibody_plant_test.cc#L153)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Just because loading SDF's is much easier. `MakeXXX()` was just the temporary hack. I clean up as I go.
</blockquote></details>

It did lead to duplicate code -- you're building the same tree with the same data with the only variation being a boolean. Still, as it hasn't happened *three* times...it gets a pass.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879#-:-LDONlFU8mA2gejBfL-T:b2lqqh8)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[multibody/multibody_tree/mobilizer_impl.h, line 88 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN3NFo3u7XgQkWrep-:-LDOTEow3b3wEekoewaL:bbls3gu) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/mobilizer_impl.h#L88)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Got it. Is there a reason that the default leaf system implementation of SetDefaultState() doesn't work?
</blockquote></details>

quaternions, of course.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 499 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDMzKFoBWXshhY3B2ua:-LDOTfTiCFuoJRpzuWMM:bip5go1) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L499)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

If you add another type of event that requires discrete update, then you have to differentiate those -- but until that happens, there's no differentiation required. And, again, the only entity that can enforce this method being called on this system is this system.

Right now I don't like the unnecessary code being here -- it suggests it is necessary and that means the code is lying.
</blockquote></details>

I'll trust you on this one. Hopefully it doesn't break my dev branch when I merge. I think you are right.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 590 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN27ek6UJtN7rv5xV1:-LDOU2X_5hoOZ-8veryB:bc1dabw) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L590)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

It should still be documented. If you can get the explanation from Evan and document it, great.  But this is an undocumented pattern and it would be a shame if it is unnecessarily propagated.
</blockquote></details>

I believe the simulator won't even call these. Though I don't see why to avoid users from calling them. A call to these should still be possible if I wanted to. Per your comment above I removed the `is_discrete` check.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/multibody_tree/multibody_plant/multibody_plant.cc, line 590 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN27ek6UJtN7rv5xV1:-LDOVP1B7LCu_4FSrETg:b-n9w0gg) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/multibody_plant/multibody_plant.cc#L590)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I believe the simulator won't even call these. Though I don't see why to avoid users from calling them. A call to these should still be possible if I wanted to. Per your comment above I removed the `is_discrete` check.
</blockquote></details>

I'll mark myself as satisfied, but I want to emphasize that the only test I was arguing was *not* necessary was in the `DoCalcDiscreteVariableUpdates()`. I don't know if these are necessary or not. I just wanted it documented as necessary or removed. So, it it explodes, it's no my fault.  :D

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879#-:-LDOVdre61ietloMNmad:bmxilld)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[multibody/multibody_tree/mobilizer_impl.h, line 88 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN3NFo3u7XgQkWrep-:-LDQ3YSF8Jhg_XA38SU6:bs2mpcg) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/mobilizer_impl.h#L88)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Well I thought you'd say that but the Declare stuff allows for a model vector that is copied by SetDefaultState (not sure if that's enough though). Did you look at that and reject it?
</blockquote></details>

No I didn't. That could probably do it, but I'm not sure, didn't try it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
@seancurtis-tri and @sherm1 I still have blocking comments from you on this one. I believe you didn't intend them to be blocking?

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879#-:-LDSBN-UASbD8eDCXk2p:b-ms0k1t)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/multibody_tree/mobilizer_impl.h, line 88 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8879#-LDN3NFo3u7XgQkWrep-:-LDU8TpHCXGsvIKqN-wt:bybcclt) ([raw file](https://github.com/robotlocomotion/drake/blob/8d6f350e6ea81e0df02e15adbb591f9a5c5b7f04/multibody/multibody_tree/mobilizer_impl.h#L88)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

No I didn't. That could probably do it, but I'm not sure, didn't try it.
</blockquote></details>

Please add a TODO mentioning that -- it would be good not to have to implement it here. (Unblocked now)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
I unblocked Sean's too since he said "I'll mark myself as satisfied".

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8879#-:-LDU8yOS-oSiepRFY64T:bldg7h3)*
<!-- Sent from Reviewable.io -->
